mogp_gpu/src/gp_gpu.hpp

```diff
@@ -0,0 +1,764 @@
+		std::string msg;
+		std::stringstream smsg(msg);
+		smsg << "All attempts at factorization failed. Last return code " << factorisation_status;
+		throw std::runtime_error(smsg.str());
+	    }
+	// for fixed nugget, add "nugget_size" to the diagonal of the matrix.
+	} else if (nugget == NUG_FIXED) {
+	    add_diagonal(n, nugget_size, dev_ptr(work_mat_d));
+
+	    factorisation_status = calc_cholesky_factors();
+	    if (factorisation_status != 0) {
+                throw std::runtime_error("Unable to factorize matrix using fixed nugget");
+	    }
+
+	} else { //nugget == "fit"
+	    factorisation_status = calc_cholesky_factors();
+	    if (factorisation_status != 0) {
+                throw std::runtime_error("Unable to factorize matrix using fitted nugget");
+	    }
+	}
+
+        identity_device(n, dev_ptr(invQ_d));
+
+        // invQ
+        status = cusolverDnDpotrs(cusolverHandle, CUBLAS_FILL_MODE_LOWER, n, n,
+                                  dev_ptr(work_mat_d), n, dev_ptr(invQ_d), n,
+                                  dev_ptr(info_d));
+
+        thrust::copy(info_d.begin(), info_d.end(), &info_h);
+        check_cusolver_status(status, info_h);
+
+
+        // invQt
+        thrust::copy(targets_d.begin(), targets_d.end(), invQt_d.begin());
+        status = cusolverDnDpotrs(cusolverHandle, CUBLAS_FILL_MODE_LOWER, n, 1,
+                                  dev_ptr(work_mat_d), n, dev_ptr(invQt_d), n,
+                                  dev_ptr(info_d));
+
+        thrust::copy(info_d.begin(), info_d.end(), &info_h);
+        check_cusolver_status(status, info_h);
+
+        // logdetC
+        thrust::device_vector<double> logdetC_d(1);
+
+        sum_log_diag(n, dev_ptr(work_mat_d), dev_ptr(logdetC_d), dev_ptr(sum_buffer_d), sum_buffer_size_bytes);
+
+        thrust::copy(logdetC_d.begin(), logdetC_d.end(), &logdetC);
+
+	//copy work_mat_d into the lower triangular Cholesky factor
+	thrust::copy(work_mat_d.begin(), work_mat_d.begin()+n*n, chol_lower_d.begin());
+
+	//set the flag to say we have fitted theta
+	theta_fitted = true;
+    }
+
+    bool get_theta_fit_status(void)
+    {
+        return theta_fitted;
+    }
+
+    void reset_theta_fit_status(void)
+    {
+        theta_fitted = false;
+    }
+
+    void get_invQ(mat_ref invQ_h)
+    {
+        thrust::copy(invQ_d.begin(), invQ_d.end(), invQ_h.data());
+    }
+
+    void get_invQt(mat_ref invQt_h)
+    {
+        thrust::copy(invQt_d.begin(), invQt_d.end(), invQt_h.data());
+    }
+
+    double get_logpost(void)
+    {
+        double result;
+        CUBLASDOT(cublasHandle, n, dev_ptr(targets_d), 1, dev_ptr(invQt_d), 1,
+                  &result);
+
+        result += logdetC + n * log(2.0 * M_PI);
+
+        return 0.5 * result;
+    }
+
+    void logpost_deriv(vec_ref result)
+    {
+        double zero(0.0);
+        double one(1.0);
+        double half(0.5);
+        double m_half(-0.5);
+
+        const int Ntheta = D + 1;
+
+        // Compute
```

Comment:
Does this correctly handle the derivative of the nugget parameter for a fitted nugget? I wasn't totally clear on how the fitted nugget is handled here.

Reply:
Previously missing, now implemented.  See above comment for details.